Add fungible token contract, via HTS

markogracin opened the pull request:
- due compatibility with Hedera (hashscan), and usage with Hashpack, we went in the HTS direction over standard ERC20.

markogracin committed:
feat(contracts): 🎸 add fungible token contract via hedera token service

markogracin committed:
feat(contracts): 🎸 bump solidity version

pcaversaccio commented on contracts/hardhat.config.ts:
I would recommend setting the optimiser settings explicitly:

```ts
    version: "0.8.28",
    settings: {
      optimizer: {
        enabled: true,
        runs: 999_999,
      },
      evmVersion: "cancun",
    },
```

Please note that Hedera is compatible with the `cancun` hard fork: https://docs.hedera.com/hedera/core-concepts/smart-contracts/deploying-smart-contracts#cancun-hard-fork. You should be aware of the semantic differences of the opcodes used on Hedera; see here: https://docs.hedera.com/hedera/core-concepts/smart-contracts/deploying-smart-contracts#solidity-variables-and-opcodes.

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.28;
+
+import './utils/HederaResponseCodes.sol';
```
Nit: please use explicit imports, e.g.:

```sol
import {HederaResponseCodes} from "./utils/HederaResponseCodes.sol";
```

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+contract TokenCreator is ExpiryHelper, HederaTokenService {
+    address private _tokenAddress;
+    address private _collateralTokenAddress;
+    address private _owner;
```
There is the full logic missing around exposing the owner publicly in the ABI & changing the owner. My recommendation would be to import this contract and inherit from it instead: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol.

markogracin replied:
Added Ownable from `@openzeppelin/contracts`, and replaced old _ownable checks, feels good!

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+    /**
+     * @dev Constructor sets the contract owner and initial fee recipient as the deployer
+     */
+    constructor() {
```
you could pass the owner address as a configurable parameter in the constructor. As proposed above, use https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol.

markogracin replied:
updated this to:

```
   constructor() Ownable(msg.sender) {
        _feeRecipient = msg.sender;
    }
```
    
During the contract deploy, I'm passing the contract owner account, but the `_feeRecipient` can be changed later on with `setFeeRecipient()`

let me know if i missed something, what is possible.. 😄

pcaversaccio replied:
lgtm like this

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+     */
+    constructor() {
+        _owner = msg.sender;
+        _feeRecipient = msg.sender;
```
```solidity
constructor(address feeRecipient_) {
    ...
    if (feeRecipient_== address(0)) {
        revert FeeReceiptInvalid(address(0));
    }
    _feeRecipient = feeRecipient_;
}
```

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+     * @notice Sets up a 1% fractional fee with min/max limits
+     * @notice Contract maintains supply and fee management permissions
+     */
+    function createFungible() external payable returns (address createdTokenAddress) {
```
I have never seen such a function tbh and I'm not sure if this the right way to do it, it's definitely not the right way on Ethereum lol

pcaversaccio replied:
Also should only be invoked by the `owner`?

markogracin replied:
Yeah, this part that differs from the standard contracts, as it's using HederaTokenService, to make it compatible with hashpack wallet and hashscan. Will find the examples and documentation that I used to add this, as a reference.

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+     * @return feeRecipient Address where fees are sent
+     * @return lockedCollateral Amount of collateral currently locked in contract
+     */
+    function getContractInfo() external view returns (
```
why not additionally defining for each variable a separate function? You can keep this function and simply invoke the functions in the `return` statement but many times you are only interested in one or two params, e.g. who is the `owner`.

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+     * @param token Address of the collateral token
+     * @return responseCode Response code from the Hedera Token Service
+     */
+    function setCollateralToken(address token) external returns (int) {
```
should be protected no? Only the `owner` should be able to do this.

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+            address(this),
+            amount
+        );
+        if (transfer != HederaResponseCodes.SUCCESS) revert("Collateral coin transfer failed");
```
This contract uses inconsistently `require` and `revert`. Can we use everywhere custom errors please?

pcaversaccio commented on contracts/contracts/TokenContract.sol:
```diff
@@ -0,0 +1,242 @@
+     * @notice Transfers all available balance to fee recipient
+     * @notice Reverts if no fees are available to collect
+     */
+    function collectFees() external {
```
we should have another emergency function to save any wrongly sent tokens to the contract.

pcaversaccio reviewed (commented):
I quickly glanced through the contract and left some comments.

markogracin commented:
> I quickly glanced through the contract and left some comments.

Thanks @pcaversaccio, appreciate it! Will take a look this weekend, and get back to re-check.

markogracin committed:
fix(contracts):🐛 add OpenZeppelin's Ownable and refactor ownership logic

Introduced OpenZeppelin's `Ownable` contract to simplify ownership management and removed custom `onlyOwner` modifier logic. Updated `TokenCreator` to inherit `Ownable`, refactored constructor, and replaced `_owner` references with `owner()`. Added `@openzeppelin/contracts` as a dependency in `package.json`, and updated import statements accordingly.

markogracin committed:
fix(contracts):🐛 update hardhat config

markogracin committed:
fix(contracts):🐛 add getter functions for key contract data

Introduce public getter functions to retrieve addresses for the token, collateral token, contract owner, and fee recipient, as well as the amount of locked collateral.

markogracin committed:
fix(contracts):🐛 restrict access to critical token functions to onlyOwner

Add `onlyOwner` modifier to `createFungible` and `setCollateralToken` functions to ensure only the contract owner can execute these actions.

markogracin committed:
fix(contracts):🐛 refactor error handling with custom errors

Replaced generic string-based revert messages with custom error types for improved clarity and gas efficiency.

markogracin committed:
feat(contracts): 🎸 add emergency token recovery function

Introduce `emergencyTokenRecovery` to allow the contract owner to recover wrongly sent tokens, ensuring collateral or main token fees remain unaffected. Added relevant event and error.
